File: kotlin/src/katas/kotlin/coroutines/steps/1.kt
@file:Suppress("EXPERIMENTAL_FEATURE_WARNING", "PackageDirectoryMismatch")

package katas.kotlin.coroutines.steps.step1

import com.natpryce.hamkrest.assertion.assertThat
import com.natpryce.hamkrest.equalTo
import kotlin.coroutines.*
import kotlin.coroutines.intrinsics.COROUTINE_SUSPENDED
import kotlin.coroutines.intrinsics.suspendCoroutineUninterceptedOrReturn

/**
 * Yield coroutine execution
 */
fun main(args: Array<String>) {
    val events = ArrayList<String>()
    fun log(it: Any?) = events.add(it.toString())

    val f = ResumableFunction.create {
        log(1)
        yield()
        log(3)
        yield()
        log(5)
    }
    log(0)
    f.resume()
    log(2)
    f.resume()
    log(4)
    f.resume()
    log(6)

    assertThat(events, equalTo(listOf("0", "1", "2", "3", "4", "5", "6")))
}

private class ResumableFunction {
    private var c: Continuation<Unit>? = null

    fun resume() {
        c?.resume(Unit)
    }

    suspend fun yield() {
        return suspendCoroutineUninterceptedOrReturn { continuation: Continuation<Unit> ->
            c = continuation
            COROUTINE_SUSPENDED
        }
    }

    companion object {
        fun create(block: suspend ResumableFunction.() -> Unit): ResumableFunction {
            val f = ResumableFunction()
            f.c = block.createCoroutine(f, completion = EmptyContinuation)
            return f
        }
    }
}

object EmptyContinuation: Continuation<Unit> {
    override val context: CoroutineContext = EmptyCoroutineContext
    override fun resumeWith(result: Result<Unit>) {}
}

File: lua/coroutines/1.lua
package.path = package.path .. ";../?.lua"
require("common")

local events = {}
function log(event)
  events[#events + 1] = event
end

-- Yield coroutine execution

local c = coroutine.create(function()
  log(2)
  coroutine.yield()
  log(4)
end)

log(1)
coroutine.resume(c)
log(3)
coroutine.resume(c)
log(5)

expect_to_be_equal(events, {
  [1] = 1,
  [2] = 2,
  [3] = 3,
  [4] = 4,
  [5] = 5,
})
